engine/src/stack.cpp

```diff
@@ -740,4 +742,80 @@ void Stack::check_completeness() const {
   }
 }
 
+[[noreturn]] void throw_missing_plugin_error(Conf conf,
+                                             std::string_view plugin_type,
+                                             std::string_view plugin_name,
+                                             const std::vector<std::string>& available_plugins) {
+  throw Error{"unknown {} plugin: {}", plugin_type, plugin_name}.explanation(
+      R"(
+              Error located in JSON segment at {}:
+
+{}
+
+              The plugin referred to by the "binding" field is not available.
+              The following {} plugins are available:
+
+                  {}
+
+              If the plugin you are looking for is not in that list, then you need
+              to get cloe-engine to load the plugin first.
+              There are several mechanisms for loading plugins:
+
+              1. Add directory to CLOE_PLUGIN_PATH environment variable:
+
+                 export CLOE_PLUGIN_PATH="$CLOE_PLUGIN_PATH:/path/to/dir/containing/plugin"
+
+              2. Add directory containing plugin from the command-line:
+
+                 cloe-engine --plugin-path=/path/to/dir/containing/plugin run config.json
+
+              3. Specify path to plugin in a stackfile directly:
+
+                 {{
+                   "version": "4",
+                   "plugins": [
+                     {{
+                       "path": "/path/to/dir/containing/plugin"
+                     }},
+                     {{
+                       "path": "/path/to/plugin.so"
+                     }}
+                   ]
+                 }}
+
+              )",
+      conf.root(), fable::indent_string(conf->dump(2), "                  "), plugin_type,
+      fable::join_vector(available_plugins, "\n                  "));
+}
+
+bool SimulatorSchema::validate(const Conf& c, std::optional<SchemaError>& err) const {
+  assert(schema_ != nullptr);
+  auto factory = c.get<std::string>(factory_key_);
```

Comment:
I'm not sure I got right what you do in the validate() function, but shouldn't you check if the schema is valid first and then check if the configured plugin is available?

Reply:
The plugin supplies the schema, so if we don't have a plugin, we don't have a schema. If the `factory_key_` is missing ("binding" in this case), we get the following error:
```
test_engine_unavailable_controller.json:/controllers/0: required property missing: binding
    In JSON segment:
        {
          "vehicle": "default"
        }
```

Reply:
This code is actually a small adaption of whats written in the base class:
```cpp
  bool validate(const Conf& c, std::optional<SchemaError>& err) const override {
    assert(schema_ != nullptr);
    auto factory = c.get<std::string>(factory_key_);
    if (!available_.count(factory)) {
      return this->set_error(err, c, "unknown factory: {}", factory);
    }

    return schema_->validate(c, err);
  }
```